addons/calendar/models/calendar.py

```diff
@@ -0,0 +1,1047 @@
+                    'time_format': res_lang.time_format}
+
+        # formats will be used for str{f,p}time() which do not support unicode in Python 2, coerce to str
+        format_date = lang_params.get("date_format", '%B-%d-%Y').encode('utf-8')
+        format_time = lang_params.get("time_format", '%I-%M %p').encode('utf-8')
+        return (format_date, format_time)
+
+    def get_display_time_tz(self, tz=False):
+        context = self.env.context.copy()
+        if tz:
+            context["tz"] = tz
+        return self.with_context(context)._get_display_time(self.start, self.stop, self.duration, self.allday)
+
+    @api.model
+    def _get_display_time(self, start, stop, zduration, zallday):
+        """
+            Return date and time (from to from) based on duration with timezone in string :
+            eg.
+            1) if user add duration for 2 hours, return : August-23-2013 at (04-30 To 06-30) (Europe/Brussels)
+            2) if event all day ,return : AllDay, July-31-2013
+        """
+        tz = self.env.context.get('tz')
+        if not tz:  # tz can have a value False, so dont do it in the default value of get !
+            tz = self.env.context.get(self.env.user.tz)
+        tz = tools.ustr(tz).encode('utf-8')  # make safe for str{p,f}time()
+
+        format_date, format_time = self.with_context(tz=tz).get_date_formats()
+        date = fields.Datetime.context_timestamp(self.with_context(tz=self.env.user.tz), fields.Datetime.from_string(start))
+        date_deadline = fields.Datetime.context_timestamp(self.with_context(tz=self.env.user.tz), fields.Datetime.from_string(stop))
+        event_date = date.strftime(format_date)
+        display_time = date.strftime(format_time)
+
+        if zallday:
+            time = _("AllDay , %s") % (event_date)
+        elif zduration < 24:
+            duration = date + timedelta(hours=zduration)
+            time = _("%s at (%s To %s) (%s)") % (event_date, display_time, duration.strftime(format_time), tz)
+        else:
+            time = _("%s at %s To\n %s at %s (%s)") % (event_date, display_time, date_deadline.strftime(format_date), date_deadline.strftime(format_time), tz)
+        return time
+
+    def _compute_is_attendee(self):
+        for meeting in self:
+            attendee = meeting._find_my_attendee()
+            meeting.is_attendee = bool(attendee)
+
+    @api.depends('final_date')
+    def _get_rulestring(self):
+        """
+        Gets Recurrence rule string according to value type RECUR of iCalendar from the values given.
+        """
+        #read these fields as SUPERUSER because if the record is private a normal search could raise an error
+        events = self.sudo().read(['id', 'byday', 'recurrency', 'final_date', 'rrule_type', 'month_by',
+                            'interval', 'count', 'end_type', 'mo', 'tu', 'we', 'th', 'fr', 'sa',
+                            'su', 'day', 'week_list'])
+        for event in events:
```

Comment:
Isn't it possible to iterate over self and prepare event dict and then call compute_rule_string with that prepared event dict and assign it to record.rrule = result of compute_rule_string(prepared event dict) ?